File: oop/sem2/lab2_2.cpp
#include <iostream>
#include <string>

using namespace std;

class Airline {
private:
    string destination, number, type, departureTime, weekDays;

public:
    Airline() {}

    Airline(string _number, string _type, string _destination, string _departureTime, string _weekDays) {
        number = _number;
        type = _type;
        destination = _destination;
        departureTime = _departureTime;
        weekDays = _weekDays;
    }

    string getDestination() {
        return destination;
    }

    bool is_weekDay(string weekDay) {
        bool is_find = false;
        for (char i : weekDays) {
            if (weekDay[0] == i) {
                is_find = true;
                break;
            }
        }
        return is_find;
    }

    bool is_timeGreater(string minDepartureTime) {
        int minHour = stoi(minDepartureTime.substr(0, 2));
        int minMinute = stoi(minDepartureTime.substr(3, 2));
        int Hour = stoi(departureTime.substr(0, 2));
        int Minute = stoi(departureTime.substr(3, 2));
        if (minHour < Hour) {
            return true;
        }
        else {
            if (minHour == Hour and minMinute < Minute ) {
                return true;
            } else {
                return false;
            }
        }
    }

    string toString() {
        return "Number: " + number + "; Type: " + type + "; Destination: " + destination +
               "; Departure time: " + departureTime + "; Week days: " + weekDays + "\n";
    }
};

void printAirlinesByDestination(Airline airlines[], int size, string destination) {
    for (int i = 0; i < size; i++) {
        if (airlines[i].getDestination() == destination) {
            cout << airlines[i].toString();
        }
    }
}

void printAirlinesByWeekDay(Airline airlines[], int size, string weekDay) {
    for (int i = 0; i < size; i++) {
        if (airlines[i].is_weekDay(weekDay)) {
            cout << airlines[i].toString();
        }
    }
}

void printAirlinesByDepartureTime(Airline airlines[], int size, string weekDay, string minDepartureTime) {
    for (int i = 0; i < size; i++) {
        if (airlines[i].is_weekDay(weekDay) & airlines[i].is_timeGreater(minDepartureTime)) {
            cout << airlines[i].toString();
        }
    }
}

int main() {
    const int n = 9;
    Airline airlines[n];
    airlines[0] = Airline("B2 727", "Boeing 737-800", "Дубай", "00:50", "4,5,6");
    airlines[1] = Airline("B2 975", "Embraer 195", "Москва", "08:45", "1,2,3,4,5,6,7");
    airlines[2] = Airline("B2 941", "Embraer 195", "Санкт-Петербург", "09:20", "1,3,4,5,6");
    airlines[3] = Airline("B2 783", "Boeing 737-800", "Стамбул", "18:25", "1,7");
    airlines[4] = Airline("B2 775", "Embraer 195", "Астана", "20:15", "1,3,5");
    airlines[5] = Airline("B2 735", "Boeing 737-800", "Тбилиси", "22:50", "3,7");
    airlines[6] = Airline("B2 737", "Embraer 195", "Батуми", "23:20", "2,4,6");
    airlines[7] = Airline("B2 123", "Airbus A380", "Нью-Йорк", "10:45", "5");
    airlines[8] = Airline("B2 346", "Boeing 767-33A", "Рим", "13:00", "1,2,3,4,5,6,7");

    string destination;
    cout << "Enter a destination: ";
    cin >> destination;
    printAirlinesByDestination(airlines, n, destination);

    string weekDay;
    cout << "\nEnter the number of the day of the week: ";
    cin >> weekDay;
    printAirlinesByWeekDay(airlines, n, weekDay);

    string minDepartureTime;
    cout << "\nEnter the minimum departure time: ";
    cin >> minDepartureTime;
    printAirlinesByDepartureTime(airlines, n, weekDay, minDepartureTime);

    return 0;
}

File: trposu/sem1/Lab2_2.java
import java.util.Scanner;

public class Lab2_2 {
    private String destination;
    private int flightNumber;
    private String typeOfPlane;
    private String departureTime;
    private String daysOfWeek;

    public Lab2_2(String destination, int flightNumber, String typeOfPlane, String departureTime, String daysOfWeek) {
        this.destination = destination;
        this.flightNumber = flightNumber;
        this.typeOfPlane = typeOfPlane;
        this.departureTime = departureTime;
        this.daysOfWeek = daysOfWeek;
    }

    public String getDestination() { return destination; }
    public String getDepartureTime() { return departureTime; }
    public String getDaysOfWeek() { return daysOfWeek; }

    public String toString() {
        return "Number: " + this.flightNumber + "; Type: " + this.typeOfPlane + "; Destination: " + destination +
                "; Departure time: " + this.departureTime + "; Week days: " + this.daysOfWeek;
    }

    public static Lab2_2[] createAirlines() {
        return new Lab2_2[]{
                new Lab2_2("New York", 101, "Boeing 747", "08:00", "Monday"),
                new Lab2_2("Los Angeles", 202, "Airbus A380", "12:00", "Tuesday, Thursday"),
                new Lab2_2("Miami", 303, "Boeing 777", "15:30", "Wednesday, Friday, Sunday"),
                new Lab2_2("Chicago", 404, "Boeing 787", "18:45", "Monday, Friday"),
                new Lab2_2("Dallas", 505, "Airbus A350", "20:15", "Tuesday, Thursday, Saturday"),
                new Lab2_2("New York", 557, "Boeing 737", "12:30", "Monday, Wednesday, Friday"),
                new Lab2_2("Los Angeles", 543, "Airbus A320", "14:45", "Monday, Thursday, Saturday"),
                new Lab2_2("Chicago", 432, "Boeing 787", "16:20", "Tuesday, Friday, Sunday"),
                new Lab2_2("Miami", 123, "Airbus A321", "18:15", "Monday, Wednesday, Friday, Sunday"),
                new Lab2_2("Seattle", 678, "Boeing 737", "20:30", "Tuesday, Thursday, Saturday"),
                new Lab2_2("Dallas", 574, "Airbus A320", "10:00", "Monday, Wednesday, Friday, Sunday"),
                new Lab2_2("San Francisco", 568, "Boeing 777", "11:45", "Tuesday, Thursday, Saturday"),
                new Lab2_2("Boston", 158, "Boeing 737", "13:10", "Monday, Wednesday, Friday, Sunday"),
                new Lab2_2("Las Vegas", 592, "Airbus A320", "15:30", "Tuesday, Thursday, Saturday"),
                new Lab2_2("Orlando", 401, "Boeing 737", "17:25", "Monday, Wednesday, Friday"),
                new Lab2_2("Houston", 406, "Airbus A321", "19:40", "Tuesday, Thursday, Saturday"),
                new Lab2_2("Denver", 193, "Boeing 737", "21:15", "Monday, Wednesday, Friday, Sunday")
        };
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Lab2_2[] airlines = Lab2_2.createAirlines();

        // Output the list of flights for a given destination
        System.out.print("Enter destination: ");
        String destination = scanner.nextLine();
        System.out.println("Flights to " + destination + ":");
        for (Lab2_2 airline : airlines) {
            if (airline.getDestination().equalsIgnoreCase(destination)) {
                System.out.println(airline);
            }
        }

        // Output the list of flights for a given day of the week
        System.out.print("\nEnter day of the week: ");
        String dayOfWeek = scanner.nextLine();
        System.out.println("Flights on " + dayOfWeek + ":");
        for (Lab2_2 airline : airlines) {
            if (airline.getDaysOfWeek().toLowerCase().contains(dayOfWeek.toLowerCase())) {
                System.out.println(airline);
            }
        }

        // Output the list of flights for a given day of the week, for which the departure time is longer than the given one
        System.out.print("\nEnter departure time (HH:MM): ");
        String departureTime = scanner.nextLine();
        System.out.println("Flights on " + dayOfWeek + " with departure time later than " + departureTime + ":");
        for (Lab2_2 airline : airlines) {
            if (airline.getDaysOfWeek().toLowerCase().contains(dayOfWeek.toLowerCase()) && airline.getDepartureTime().compareTo(departureTime) > 0) {
                System.out.println(airline);
            }
        }
    }
}
